feat(keycard): keycard details are being synced among devices

saledjenic opened the pull request:
Keycard details are being synced among devices:
- sync all keycards state
- sync every keycard change

Entire work is splitted across few independent commits.

How it fits the desktop app can be seen in the video in this pr https://github.com/status-im/status-desktop/pull/9600

saledjenic committed:
chore(keycard): keypair tests moved to a new separate test file

saledjenic committed:
chore(keycard): `AddMigratedKeyPair` renamed to `AddMigratedKeyPairOrAddAccountsIfKeyPairIsAdded` which better describes what that function does

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (52)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | b0b9f780 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/1/) | 2023-02-21 08:15:35 | ~23 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/1/consoleText) |
| :heavy_check_mark: | b0b9f780 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/1/) | 2023-02-21 08:16:37 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230221-081512-b0b9f7-pr3211.zip) |
| :heavy_check_mark: | b0b9f780 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/1/) | 2023-02-21 08:19:32 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230221-081512-b0b9f7-pr3211.aar) |
| :heavy_multiplication_x: | b0b9f780 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/1/) | 2023-02-21 08:22:11 | ~6 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/1//consoleText) |
| :heavy_multiplication_x: | b0b9f780 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/2/) | 2023-02-21 08:49:06 | ~22 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/2//consoleText) |
| | | | | | | |
| :x: | 9e52f522 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/2/) | 2023-02-21 11:41:02 | ~12 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/2/consoleText) |
| :heavy_check_mark: | 9e52f522 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/2/) | 2023-02-21 11:41:45 | ~55 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230221-114049-9e52f5-pr3211.zip) |
| :heavy_check_mark: | 9e52f522 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/2/) | 2023-02-21 11:44:50 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230221-114049-9e52f5-pr3211.aar) |
| :heavy_multiplication_x: | 9e52f522 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/3/) | 2023-02-21 12:02:27 | ~21 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 7b007a64 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/3/) | 2023-02-22 15:11:57 | ~55 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230222-151101-7b007a-pr3211.zip) |
| :heavy_multiplication_x: | 7b007a64 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/4/) | 2023-02-22 15:14:34 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/4//consoleText) |
| :heavy_check_mark: | 7b007a64 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/3/) | 2023-02-22 15:14:42 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230222-151101-7b007a-pr3211.zip) |
| :heavy_check_mark: | 7b007a64 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/3/) | 2023-02-22 15:15:47 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230222-151101-7b007a-pr3211.aar) |
| | | | | | | |
| :heavy_multiplication_x: | cc121506 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/5/) | 2023-02-22 15:15:46 | ~59 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/5//consoleText) |
| :heavy_check_mark: | cc121506 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/4/) | 2023-02-22 15:16:01 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230222-151446-cc1215-pr3211.zip) |
| :heavy_check_mark: | cc121506 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/4/) | 2023-02-22 15:17:00 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230222-151444-cc1215-pr3211.zip) |
| :heavy_check_mark: | cc121506 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/4/) | 2023-02-22 15:19:25 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230222-151550-cc1215-pr3211.aar) |
| | | | | | | |
| :heavy_check_mark: | f5b3d4f0 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/5/) | 2023-02-22 15:17:31 | ~50 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230222-151641-f5b3d4-pr3211.zip) |
| :heavy_multiplication_x: | f5b3d4f0 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/6/) | 2023-02-22 15:17:33 | ~51 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/6//consoleText) |
| :heavy_check_mark: | f5b3d4f0 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/5/) | 2023-02-22 15:19:10 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230222-151703-f5b3d4-pr3211.zip) |
| :heavy_check_mark: | f5b3d4f0 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/5/) | 2023-02-22 15:22:36 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230222-151928-f5b3d4-pr3211.aar) |
| | | | | | | |
| :heavy_check_mark: | f96e45d5 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/6/) | 2023-02-22 15:34:42 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230222-153316-f96e45-pr3211.zip) |
| :heavy_check_mark: | f96e45d5 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/6/) | 2023-02-22 15:35:29 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230222-153316-f96e45-pr3211.zip) |
| :heavy_check_mark: | f96e45d5 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/6/) | 2023-02-22 15:36:45 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230222-153316-f96e45-pr3211.aar) |
| :heavy_check_mark: | f96e45d5 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/7/) | 2023-02-22 15:44:24 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/7//consoleText) |
| | | | | | | |
| :heavy_check_mark: | a14fa3ae | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/7/) | 2023-02-22 21:01:29 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230222-210028-a14fa3-pr3211.zip) |
| :heavy_check_mark: | a14fa3ae | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/7/) | 2023-02-22 21:02:53 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230222-210028-a14fa3-pr3211.zip) |
| :heavy_check_mark: | a14fa3ae | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/7/) | 2023-02-22 21:03:43 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230222-210028-a14fa3-pr3211.aar) |
| :heavy_multiplication_x: | a14fa3ae | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/8/) | 2023-02-22 21:23:20 | ~22 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/8//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 0eedd4e0 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/8/) | 2023-02-22 22:00:37 | ~50 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230222-215946-0eedd4-pr3211.zip) |
| :heavy_check_mark: | 0eedd4e0 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/8/) | 2023-02-22 22:02:05 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230222-215946-0eedd4-pr3211.zip) |
| :heavy_check_mark: | 0eedd4e0 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/8/) | 2023-02-22 22:03:11 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230222-215946-0eedd4-pr3211.aar) |
| :heavy_check_mark: | 0eedd4e0 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/9/) | 2023-02-22 22:10:12 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/9//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 52bb64fd | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/9/) | 2023-02-23 11:17:45 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230223-111628-52bb64-pr3211.zip) |
| :heavy_check_mark: | 52bb64fd | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/9/) | 2023-02-23 11:19:51 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230223-111623-52bb64-pr3211.aar) |
| :heavy_check_mark: | 52bb64fd | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/9/) | 2023-02-23 11:20:24 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230223-111628-52bb64-pr3211.zip) |
| :heavy_check_mark: | 52bb64fd | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/10/) | 2023-02-23 11:28:16 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/10//consoleText) |
| | | | | | | |
| :heavy_check_mark: | fadfd109 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/10/) | 2023-02-23 11:19:02 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230223-111758-fadfd1-pr3211.zip) |
| :heavy_check_mark: | fadfd109 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/10/) | 2023-02-23 11:23:08 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230223-112028-fadfd1-pr3211.zip) |
| :heavy_check_mark: | fadfd109 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/10/) | 2023-02-23 11:23:11 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230223-111955-fadfd1-pr3211.aar) |
| | | | | | | |
| :heavy_check_mark: | e7674550 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/11/) | 2023-02-23 11:27:10 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230223-112604-e76745-pr3211.zip) |
| :heavy_check_mark: | e7674550 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/11/) | 2023-02-23 11:29:04 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230223-112604-e76745-pr3211.zip) |
| :heavy_check_mark: | e7674550 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/11/) | 2023-02-23 11:29:56 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230223-112600-e76745-pr3211.aar) |
| :heavy_check_mark: | e7674550 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/11/) | 2023-02-23 11:39:09 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/11//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 66373d4e | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/12/) | 2023-02-23 14:59:56 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230223-145839-66373d-pr3211.zip) |
| :heavy_check_mark: | 66373d4e | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/12/) | 2023-02-23 15:02:14 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230223-145835-66373d-pr3211.aar) |
| :heavy_check_mark: | 66373d4e | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/12/) | 2023-02-23 15:02:16 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230223-145839-66373d-pr3211.zip) |
| :heavy_check_mark: | 66373d4e | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/12/) | 2023-02-23 15:21:24 | ~22 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/12//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 6f197df1 | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/13/) | 2023-02-24 13:13:16 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230224-131211-6f197d-pr3211.zip) |
| :heavy_multiplication_x: | 6f197df1 | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/13/) | 2023-02-24 13:14:24 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/13//consoleText) |
| :heavy_check_mark: | 6f197df1 | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/13/) | 2023-02-24 13:16:31 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230224-131209-6f197d-pr3211.aar) |
| :heavy_check_mark: | 6f197df1 | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/13/) | 2023-02-24 13:20:46 | ~8 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230224-131210-6f197d-pr3211.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | df0a2a4d | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/14/) | 2023-02-27 08:25:25 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230227-082418-df0a2a-pr3211.zip) |
| :heavy_check_mark: | df0a2a4d | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/14/) | 2023-02-27 08:28:15 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230227-082416-df0a2a-pr3211.aar) |
| :heavy_check_mark: | df0a2a4d | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/14/) | 2023-02-27 08:30:34 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230227-082416-df0a2a-pr3211.zip) |
| :heavy_check_mark: | df0a2a4d | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/14/) | 2023-02-27 08:34:34 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/14//consoleText) |
| | | | | | | |
| :heavy_check_mark: | fc0d096f | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3211/15/) | 2023-02-27 14:41:27 | ~56 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230227-144031-fc0d09-pr3211.zip) |
| :heavy_check_mark: | fc0d096f | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3211/15/) | 2023-02-27 14:43:23 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230227-144031-fc0d09-pr3211.zip) |
| :heavy_check_mark: | fc0d096f | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3211/15/) | 2023-02-27 14:44:14 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230227-144026-fc0d09-pr3211.aar) |
| :heavy_check_mark: | fc0d096f | [#15](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/15/) | 2023-02-27 15:01:21 | ~20 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3211/15//consoleText) |

saledjenic committed:
chore(keycard): `last_update_clock` column added to `keycards` table

`last_update_clock` will be used later for synchronization.
All keypair functions take clock value in consideration when
making a decision whether to perform an action or not.

saledjenic committed:
feat(keycard): keycard details are being synced among devices

- sync all keycards state
- sync every keycard change

richard-ramos commented on protocol/protobuf/pairing.proto:
```diff
@@ -298,3 +298,34 @@ message SyncRawMessage {
+  bool locked = 3;
+  string key_uid = 4;
+  repeated bytes addresses = 5;
+  uint64 last_update_clock = 6;
+}
+
+message SyncKeycardAction {
+  Action action = 1;
+  string oldKeycardUid = 2;
+  SyncKeycard keycard = 3;
+
+  enum Action {
+    KEYCARD_ADDED = 0;
+    ACCOUNTS_ADDED = 1;
+    KEYCARD_DELETED = 2;
+    ACCOUNTS_REMOVED = 3;
+    LOCKED = 4;
+    UNLOCKED = 5;
+    UID_UPDATED = 6;
+    NAME_CHANGED = 7;
+  }
+}
+
+message SyncAllKeycards {
+  repeated SyncKeycard keycards = 1;
+  uint64 syncing_clock = 2;
```
This is me being nitpicky, so comment could be ignored, but could the `*_clock` attributes be named `clock`, similar to how it's done in other .proto files?

richard-ramos commented on multiaccounts/keypairs/database.go:
```diff
@@ -14,6 +20,48 @@ type KeyPair struct {
+	if index < 0 || index >= len(s) {
+		panic("keycard: index out of the range")
+	}
+	return append(s[:index], s[index+1:]...)
```
nice!

richard-ramos commented on multiaccounts/keypairs/database.go:
```diff
@@ -2,10 +2,16 @@ package keypairs
+	"github.com/status-im/status-go/protocol/protobuf"
+)
+
+const (
+	dbTransactionIsNil = "keycard: database transaction is nil"
 )
```
```suggestion
var dbTransactionIsNil = errors.New("keycard: database transaction is nil")
```

richard-ramos reviewed (approved):
Cool PR :crown:

saledjenic commented:
@richard-ramos thanks for the review. I addressed all your comments.

qfrank commented on protocol/messenger.go:
```diff
@@ -2348,6 +2348,12 @@ func (m *Messenger) SyncDevices(ctx context.Context, ensName, photoPath string,
 	}
+
+	err = m.syncAllKeycards(ctx, rawMessageHandler)
+	if err != nil {
```
Hi @saledjenic , can you elebrate why only syncAllKeycards which corresponding protobuf.SyncAllKeycards within SyncDevices? we don't take care of protobuf.SyncKeycardAction within SyncDevices?

saledjenic replied:
@qfrank yes, sure... this is how UI looks in the desktop app:

`SyncKeycard` message is used to send details about each keycard change to enabled paired devices so they can update the their state for changed keycard.

But if two devices are using the same seed phrase and they are not paired, or pairing was disabled at the moment when some change were made on one of devices, user is able to sync the enabled paired devices clicking on `Sync All Devices` button (image below), in that case we're sending `SyncAllKeycards` message which contains the list of `SyncKeycard` messages that the paired devices can update their states based on that list. Cause in that case the device on which `Sync All Devices` button was clicked becomes source of truth.

![qq](https://user-images.githubusercontent.com/86303051/221168932-1d65d074-d9da-4421-b6fc-df12bfb82a1d.png)

qfrank replied:
sorry, my question maybe not clear, i mean why we don't process syncing/dispatch protobuf.SyncKeycardAction within SyncDevices so that another device will receive message  which's type is protobuf.SyncKeycardAction? I knew syncAllKeycards will dispatch message protobuf.SyncKeycard and i have no question about it :) @saledjenic

saledjenic replied:
@qfrank we do that on each keycard action.

Please check the `dispatchKeycardActivity` function here https://github.com/status-im/status-go/blob/66373d4e0201796222026ce263948c0069be3807/protocol/messenger_keycard.go#L87-L113

Check this [video](https://drive.google.com/file/d/1FnOUlld0iSZPrDz-eGqyrbzfnlXguPrr/view?usp=share_link) you will see that each action is dispatched and changes are received and applied on other device.

qfrank commented on protocol/messenger_sync_raw_messages.go:
```diff
@@ -188,6 +188,28 @@ func (m *Messenger) HandleSyncRawMessages(rawMessages []*protobuf.RawMessage) er
+				m.logger.Error("failed to handleSyncKeycards when HandleSyncRawMessages", zap.Error(err))
+				continue
+			}
+		case protobuf.ApplicationMetadataMessage_SYNC_KEYCARD_ACTION:
```
if this type of message(ApplicationMetadataMessage_SYNC_KEYCARD_ACTION) doesn't come from SyncDevices, then i think we don't need process it here, since method HandleSyncRawMessages only invoked during local pairing @saledjenic

saledjenic replied:
@qfrank you're perfectly right, thanks for the heads-up, updated that.
